src/jit/codegencommon.cpp

```diff
@@ -506,9 +506,7 @@ void CodeGenInterface::genUpdateRegLife(const LclVarDsc* varDsc, bool isBorn, bo
 }
 
 //----------------------------------------------------------------------
-// compNoGCHelperCallKillSet:
-//
-// Gets a register mask that represents the kill set for a helper call.
+// compHelperCallKillSet: Gets a register mask that represents the kill set for a helper call.
```

Comment:
Look like this function: `compHelperCallKillSet` has the same meaning as `compNoGCHelperCallKillSet`, but the first is used during `LSRA` and the second during the emitting. `compNoGCHelperCallKillSet` is also more conservative and returns bigger kill sets for the same helper.

Was this separation intentional? Do we need to merge them?

Reply:
They serve somewhat different purposes.

The `compHelperCallKillSet` is the classic kill set for liveness purposes -- it returns the set of registers that might be modified by the call. Note a live-across GC reference in a register may change bit values during the call if a GC happens during the call and relocates the referred-to object, but the new bit pattern is still the same reference for classical liveness purposes.

The `compNOGCHelperCallKillSet` returns the subset of killed registers that no longer can contain GC references after the call. So a register that is modified by the call and still contains a GC reference after the call will be reported live by this method and dead by the method above. 

`compNOGCHelperCallKillSet` should generally return a smaller kill set than `compHelperCallKillSet`.

Usually the modifications to the GC references in calls that are special-cased by `compNOGCHelperCallKillSet` are done in some predictable way (eg the references are byrefs and are incremented) so they are still useful to the jit after the call.

You can alternatively think of helpers that are special-cased by `compNOGCHelperCallKillSet` as helpers that return multiple results via those extra live registers. When those extra results are GC refs then they need to be reported as live for GC, even if classically they are killed.

Reply:
Thank you, it becomes clearer now.

> compNOGCHelperCallKillSet should generally return a smaller kill set than compHelperCallKillSet.

I think right now we often do the opposite, for example, `CORINFO_HELP_PROF_FCN_LEAVE` on arm returns empty set from `compHelperCallKillSet` and `RBM_CALLEE_TRASH_NOGC & ~RBM_PROFILER_RET_SCRATCH` from `compNoGCHelperCallKillSet`.

I think it should be possible to check that `compNoGCHelperCallKillSet(helper) & compHelperCallKillSet(helper) == compNoGCHelperCallKillSet(helper)`, but now it will fail in many cases, do we want to fix it?

Reply:
We might want to fix it (the subset issue); you might want to enumerate the cases where it's wrong currently.